spec/migrations/20200607025146_add_ancestry_to_vm_spec.rb

```diff
@@ -0,0 +1,182 @@
+      #    e   f
+      it 'updates ancestry' do
+        tree = create_tree(:a => [{:c => :g}, {:b => {:d => [:e, :f]}}])
+        a, b, c, d, e, f, g = tree[:a], tree[:b], tree[:c], tree[:d], tree[:e], tree[:f], tree[:g]
+
+        migrate
+
+        expect(find_rel(a).ancestry).to eq(nil)
+        expect(find_rel(b).ancestry).to eq(ancestry_for(find_rel(a)))
+        expect(find_rel(c).ancestry).to eq(ancestry_for(find_rel(a)))
+        expect(find_rel(g).ancestry).to eq(ancestry_for(find_rel(c), find_rel(a)))
+        expect(find_rel(e).ancestry).to eq(ancestry_for(find_rel(d), find_rel(b), find_rel(a)))
+        expect(find_rel(f).ancestry).to eq(ancestry_for(find_rel(d), find_rel(b), find_rel(a)))
+      end
+
+      it 'does not change ems_metadata tree' do
+        tree = create_tree(:a => [{:c => :g}, {:b => {:d => [:e, :f]}}])
+        a, b, c, d, e, f, g = tree[:a], tree[:b], tree[:c], tree[:d], tree[:e], tree[:f], tree[:g]
+        create_non_genealogy_rel(ancestry_for(a), c)
+        create_non_genealogy_rel(ancestry_for(a, c), g)
+        create_non_genealogy_rel(ancestry_for(a), b)
+        create_non_genealogy_rel(ancestry_for(b, a), d)
+        create_non_genealogy_rel(ancestry_for(d, b, a), e)
+        create_non_genealogy_rel(ancestry_for(d, b, a), f)
+        create_non_genealogy_rel(nil, a)
+
+        migrate
+
+        expect(find_rel(a).ancestry).to eq(nil)
+        expect(find_rel(b).ancestry).to eq(ancestry_for(find_rel(a)))
+        expect(find_rel(c).ancestry).to eq(ancestry_for(find_rel(a)))
+        expect(find_rel(g).ancestry).to eq(ancestry_for(find_rel(c), find_rel(a)))
+        expect(find_rel(e).ancestry).to eq(ancestry_for(find_rel(d), find_rel(b), find_rel(a)))
+        expect(find_rel(f).ancestry).to eq(ancestry_for(find_rel(d), find_rel(b), find_rel(a)))
+        expect(rel_stub.count).to eq(14)
+        expect(rel_stub.all.pluck(:relationship).uniq).to contain_exactly("genealogy", "ems_metadata")
+      end
+    end
+  end
+
+  private
+
+  def find_rel(obj)
+    rel_stub.all.detect { |r| r.resource_id == obj.id }
+  end
+
+  def create_tree(tree, relationship = default_rel_type)
+    resources = {}
+    traverse(tree, []) { |_, id| resources.merge!(id => vm_stub.create!) }
+
+    traverse(tree, []) do |parents, id|
+      ancestry = parents.reverse.map { |s| rel_stub.find_by(:resource_id => resources[s].id).id }.compact.join('/') if parents.present?
+      rel_stub.create!(:ancestry => ancestry, :resource_id => resources[id].id, :resource_type => 'VmOrTemplate', :relationship => relationship)
+    end
+
+    resources
+  end
+
+  def traverse(tree, parent, &block)
+    case tree
+    when Symbol || String
+      yield(parent, tree)
+    when Array
+      tree.each { |node| traverse(node, parent, &block) }
+    when Hash
+      tree.each do |key, children|
+        yield(parent, key)
+        traverse(children, parent + [key], &block)
+      end
+    else
+      puts "curious type: #{tree.class.name}"
```

Comment:
Minor, but should this raise an Exception instead?  Either way, this is a debugging statement so it should probably be removed as per the bot's request.